### PR TITLE
fix(governance): some governance invitations use proposal patterns

### PR DESCRIPTION
**Pull request by erights**
Start replacing uses of the deprecated `assertProposalShape` with `makeInvitation`s direct support for proposalShape patterns. This provides better protection for contracts against malformed offers. When an offer's proposal violates the contract's required proposalShape pattern, it is rejected in Zoe and the contract never even sees the malformed offer.

This PR is somewhat empirical, revising the provided proposal patterns so that they were tighter than the `assertProposalPattern` call it replaced, while still passing all tests. For each modified contract, it would be good for someone who knows the meaning of the contract to do a sanity check on the proposalShape pattern.

One particular way these patterns could be tighter, to protect the contract from seeing more malformed offers, is to use the amountShape specific to the brand the contract is willing to accept in that position. I skipped that additional work, leaving a TODO there instead. Reviewers, if you know what brand-specific amountShape I should use in any such case, please let me know.

# Upgrade considerations

We definitely need to think hard about whether this PR introduces any representation changes that create an upgrade issue. The proposal shapes that we're less worried about were extracted into https://github.com/Agoric/agoric-sdk/pull/8268 , leaving only some governance cases here.

### CODE REVIEW

File: packages/governance/src/committee.js
Comment (dckc) [inline packages/governance/src/committee.js]: :+1:

File: packages/governance/src/committee.js
Comment (dckc) [inline packages/governance/src/committee.js:90]: :+1:

File: packages/governance/src/electorateTools.js
Comment (dckc) [inline packages/governance/src/electorateTools.js:103]: :+1:

File: packages/governance/src/noActionElectorate.js
Comment (dckc) [inline packages/governance/src/noActionElectorate.js:45]: :+1:

File: packages/inter-protocol/src/stakeFactory/attestation.js
Comment (dckc) [inline packages/inter-protocol/src/stakeFactory/attestation.js]: :+1:

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js
Comment (dckc) [inline packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js]: I think it comes from...

https://github.com/Agoric/agoric-sdk/blob/ef35a91f56d47314f2bed14e7c45c3f319203850/packages/inter-protocol/src/stakeFactory/stakeFactory.js#L201

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js
Comment (dckc) [inline packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js]: So the TODO above is critical; we shouldn't land this delete without also landing the pattern.

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js
Comment (dckc) [inline packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js]: I don't think so... but only because I don't see a case where you would want to give both collateral and debt-brand, or want both. Why is that allowed in vaultFactory?

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js
Comment (dckc) [inline packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js]: :+1:

File: packages/pegasus/src/pegasus.js
Comment (dckc) [inline packages/pegasus/src/pegasus.js]: :+1:

**Review by dckc — COMMENTED** [review]
I commented on the ones I know.

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js
Comment (turadg) [reply packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js]: The requirement for VaultFactory AdjustBalances is that you can "give" or "want" without specifying the other. 

I don't know either of a good use case for giving the same brands as you're wanting, but there's also not a strong reason to forbid that and allowing it makes the shape simpler to implement the above requirement.

Doesn't the StakeFactory have the same requirement? i.e. that you can specify one and the other is implied by your staked holdings. If so we lack test coverage for that case because this shape isn't throwing.

File: packages/inter-protocol/src/vaultFactory/liquidateIncrementally.js
Comment (turadg) [inline packages/inter-protocol/src/vaultFactory/liquidateIncrementally.js]: I first read this as some sort of parameter to AmountShape. Could be more explicit:
```suggestion
      In: AmountShape, // TODO get amount shape from brand
```

When do you think that should be done?

File: packages/inter-protocol/src/vaultFactory/vault.js
Comment (turadg) [inline packages/inter-protocol/src/vaultFactory/vault.js]:
```suggestion
  // Adjustments must be able to specify just the 'give' or the 'want'.
  // There's no use case for giving both brands at once, but it allows
  // for the above requirement and is supported by the contract
  const AdjustBalancesProposalShape = M.splitRecord({
```

File: packages/inter-protocol/src/vaultFactory/vault.js
Comment (turadg) [inline packages/inter-protocol/src/vaultFactory/vault.js]: This asserts something that can't be known at interface guard time: proposal requirements given the state of the vault.

Without it, `given` on line 458 will be undefined which will cause this error,
```
"leftAmount" "[undefined]" must be a pass-by-copy record, not "undefined
```
That's a UX regression from (approx):
```
actual undefined did not match expected AmountShape
```
But that's not perfect either and if we really want to eliminate assertProposalShape then this shouldn't use it anyway.

Please add after `getProposal()`,
```js
given || Fail`closing an active vault requires a give`
```

File: packages/inter-protocol/src/vaultFactory/vaultManager.js
Comment (turadg) [inline packages/inter-protocol/src/vaultFactory/vaultManager.js]: 👍

File: packages/inter-protocol/src/vpool-xyk-amm/addLiquidity.js
Comment (turadg) [inline packages/inter-protocol/src/vpool-xyk-amm/addLiquidity.js]: Everything in `vpool-xyk-amm` can be omitted because there's an open PR to delete. I'll skip review of these.

File: packages/pegasus/src/pegasus.js
Comment (turadg) [inline packages/pegasus/src/pegasus.js]: consider normalizing casing,
```suggestion
const TransferProposalShape = M.splitRecord({
```

File: packages/zoe/src/contractSupport/zoeHelpers.js
Comment (turadg) [inline packages/zoe/src/contractSupport/zoeHelpers.js]: doesn't it have to be an `AmountKeywordRecord`?
```suggestion
      give: AmountKeywordRecordShape,
```

File: packages/zoe/src/contracts/auction/index.js
Comment (turadg) [inline packages/zoe/src/contracts/auction/index.js]: wondering whether it's always worth defining these constants that will be used in exactly one place.

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js
Comment (Chris-Hibbert) [reply packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js]: I can't tell whether this is redundant with what @turadg said, but I want to be explicit:

with VF AdjustBalances, you can give both or want both or give one and want the other. I don't think it ever makes sense to have the same keyword in both give and want.

File: packages/inter-protocol/src/vaultFactory/vault.js
Comment (Chris-Hibbert) [reply packages/inter-protocol/src/vaultFactory/vault.js]: There is a use case for giving both at once. If you want to rescue a vault that's on the verge of being liquidated, and you have limited resources, you might add collateral at the same time that you're repaying IST.

File: packages/zoe/src/contracts/auction/index.js
Comment (dckc) [reply packages/zoe/src/contracts/auction/index.js]: They might be handy for clients to use. (recall our discussion of `client-utils.js` things to accompany contracts).

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js
Comment (dckc) [reply packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js]: > If so we lack test coverage for that case ...

That's where my mind went too.

I'm not sure I can justify the time to swap in the relevant context for this just now.

File: packages/zoe/src/contracts/auction/index.js
Comment (erights) [reply packages/zoe/src/contracts/auction/index.js]: > They might be handy for clients to use. (recall our discussion of client-utils.js things to accompany contracts).

What discussion is that? Sounds interesting.

File: packages/inter-protocol/src/vaultFactory/liquidateIncrementally.js
Comment (erights) [reply packages/inter-protocol/src/vaultFactory/liquidateIncrementally.js]: Done.

> When do you think that should be done?

No strong opinion. If you have any to suggest for this PR, that'd be great. But even with these left for later, this PR is still strictly stricter than the status quo.

When do you think that should be done?

File: packages/zoe/src/contracts/auction/index.js
Comment (dckc) [reply packages/zoe/src/contracts/auction/index.js]: Things like

 - utilities to do AMM calculations in a web UI
 - contract keyword constants or helper functions for constructing offers?
 - governed parameter utilities (several contracts have a `params.js` file that's useful for callers of `startInstance`)

Comment (turadg) [reply packages/zoe/src/contracts/auction/index.js]: https://github.com/Agoric/agoric-sdk/pull/7091

Though note that the production shapes won't work for "remotables" in clientSupport because they use BoardRemotes. Maybe eventually we'll want to get rid of BoardRemotes and leave that to the marshaller.

File: packages/inter-protocol/src/vaultFactory/liquidateIncrementally.js
Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/liquidateIncrementally.js]: probably after Vaults Release. though I might slip it into making vaultFactory upgradeable. (might help the factoring to bear this requirement in mind)

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js
Comment (erights) [reply packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js]: Good. Added a proposalShape there and deleted the assertProposalShape here. Done.

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js
Comment (erights) [reply packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js]: Yes. And I mistakenly deleted this one without adding a corresponding proposalShape to the `makeInvitation` for `closeHook`. Fixed.

File: packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js
Comment (erights) [reply packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js]: The goal for this PR is to express with patterns at least the restrictions that were expressed by `assertProposalShape` and similar, but still be unrestrictive enough to allow everything we should be allowing. So, based on what I think I understood of the above comments, I did revise this pattern to be like the `AdjustBalancesProposalShape` from VF, that is more accepting.

PTAL and let me know if I misunderstood. Leaving this Conversation unresolved until I hear back.

File: packages/inter-protocol/src/stakeFactory/stakeFactory.js
Comment (erights) [inline packages/inter-protocol/src/stakeFactory/stakeFactory.js]: Seems annoyingly similar to the `zcf.makeInvitation` call in `stakeFactoryKit` function `makeCloseInvitation`. As a result, the proposalShape used here is annoyingly like the one used there. Can we consolidate both?

Leaving Conversation unresolved until I hear back.

File: packages/inter-protocol/src/vaultFactory/vault.js
Comment (erights) [reply packages/inter-protocol/src/vaultFactory/vault.js]: In light of @Chris-Hibbert 's comment and @turadg thumbs-up agreement, kept just the first line of the comment. Also added that line to the `AdjustBalancesProposalShape` which is now similar to this one.

Done.

File: packages/inter-protocol/src/vaultFactory/vault.js
Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/vault.js]: @erights I see new commits since this request. Not sure if you disagree or just haven't noticed it.

File: packages/inter-protocol/src/vaultFactory/liquidateIncrementally.js
Comment (turadg) [inline packages/inter-protocol/src/vaultFactory/liquidateIncrementally.js]: I believe the liquidation files are going away too so won't review this.

File: packages/inter-protocol/src/vpool-xyk-amm/removeLiquidity.js
Comment (turadg) [inline packages/inter-protocol/src/vpool-xyk-amm/removeLiquidity.js]: non blocking but please consider leaving alone vpool-xyk-amm since it's being removed. i.e. restore master

File: packages/inter-protocol/src/vaultFactory/vault.js
Comment (erights) [reply packages/inter-protocol/src/vaultFactory/vault.js]: Just haven't gotten to it yet.

File: packages/inter-protocol/src/vpool-xyk-amm/addLiquidity.js
Comment (erights) [reply packages/inter-protocol/src/vpool-xyk-amm/addLiquidity.js]: @Chris-Hibbert 's deletion PR beat me to it ;)

Gone.

File: packages/inter-protocol/src/vpool-xyk-amm/removeLiquidity.js
Comment (erights) [reply packages/inter-protocol/src/vpool-xyk-amm/removeLiquidity.js]: Gone

File: packages/inter-protocol/src/vaultFactory/vault.js
Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/vault.js]: Ok, I did it here: https://github.com/Agoric/agoric-sdk/pull/7094

That's my take on the vaultFactory changes in this PR and the other changes have other reviewers so I've unsubscribed here.

File: packages/inter-protocol/src/stakeFactory/stakeFactory.js
Comment (Chris-Hibbert) [reply packages/inter-protocol/src/stakeFactory/stakeFactory.js]: seems like they could be unified. What's the obstacle? The factory has the branded amounts. Can the kit get them from there? Is there an obstacle to the Kit making the branded amounts itself?

File: packages/inter-protocol/src/vaultFactory/vault.js
Comment (erights) [reply packages/inter-protocol/src/vaultFactory/vault.js]: Thx. Reverted change here in favor of #7094. Resolving.

File: packages/inter-protocol/src/vaultFactory/liquidateIncrementally.js
Comment (erights) [reply packages/inter-protocol/src/vaultFactory/liquidateIncrementally.js]: In light of that, I reverted this file. Resolving.

File: packages/zoe/src/contracts/auction/index.js
Comment (erights) [reply packages/zoe/src/contracts/auction/index.js]: I leaving as separate constant for now. Resolving.

File: packages/inter-protocol/src/stakeFactory/stakeFactory.js
Comment (erights) [reply packages/inter-protocol/src/stakeFactory/stakeFactory.js]: @Chris-Hibbert I recall we discussed this, but I no longer remember where we landed. Do you?

Comment (Chris-Hibbert) [reply packages/inter-protocol/src/stakeFactory/stakeFactory.js]: I don't remember the conversation. Shall we try again?

**Comment by erights** [comment]
We definitely need to think hard about whether this PR introduces any representation changes that create an upgrade issue.

**Comment by erights** [comment]
Reviewers,

Enough has changed that it is worth asking afresh: What's needed for me to move this forward? Because
   * this contributes to defensiveness against incoming messages, 
   * may be easy since it has stayed green for a long time
   * makes more of our code a better example for others to follow
 
I would like to make progress on this. But no great urgency.

**Comment by erights** [comment]
If some cases remain problematic, I'd rather just postpone those to a later PR and make progress now on the non-problematic ones. Thanks.

**Comment by dckc** [comment]
> What's needed for me to move this forward?

The emerging norm is to have [upgrade tests](https://github.com/Agoric/agoric-sdk/tree/master/packages/deployment/upgrade-test) for any contracts deployed in production; in particular, `committee.js`.

Pegasus is not (yet) in production, so landing that seems un-problematic. I think `auction.js` likewise.

For the helper functions, I'd like to see them traced back to any production bundles that they might appear in.

**Review by dckc — COMMENTED** [review]
I'm not sure about landing changes to committee.js without at least a scheduled issue for upgrade testing. @turadg ?

**Comment by turadg** [comment]
> I'm not sure about landing changes to committee.js without at least a scheduled issue for upgrade testing. @turadg ?

Given the low urgency of landing this and that we have upgrade-12 scheduled, I'd request such a test be included in this PR before considering landing it.

>  We definitely need to think hard about whether this PR introduces any representation changes that create an upgrade issue.

Please include this in an "Upgrade considerations" section in the main PR description.

**Comment by erights** [comment]
> Please include this in an "Upgrade considerations" section in the main PR description.

Done

**Comment by erights** [comment]
> > I'm not sure about landing changes to committee.js without at least a scheduled issue for upgrade testing. @turadg ?
>
> Given the low urgency of landing this and that we have upgrade-12 scheduled, I'd request such a test be included in this PR before considering landing it.

What if I postponed the committee.js changes to a later PR? What if I postponed all the packages/governance changes to a later PR?

**Comment by dckc** [comment]
Separating governance might do it. Would have to check whether the zoe changes affect any production bundles too.

**Comment by erights** [comment]
Zoe and Pegasus cases extracted to https://github.com/Agoric/agoric-sdk/pull/8268 , leaving only the governance cases here.

**Review by Chris-Hibbert — APPROVED** [review]
Looks good.

**Review by turadg — APPROVED** [review]

**Pull request merged**
